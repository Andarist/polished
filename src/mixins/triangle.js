// @flow
import messageHandlers, {
  typeChecks,
} from '../internalHelpers/_messageHandlers'

/** */
type PointingDirection = 'top' | 'right' | 'bottom' | 'left'

type TriangleArgs = {
  backgroundColor?: string,
  foregroundColor: string,
  height: string,
  width: string,
  pointingDirection: PointingDirection,
}

const getBorderWidth = (
  pointingDirection: PointingDirection,
  height: number,
  width: number,
) => {
  switch (pointingDirection) {
    case 'top':
      return `0 ${width / 2}px ${height}px ${width / 2}px`
    case 'left':
      return `${height / 2}px ${width}px ${height / 2}px 0`
    case 'bottom':
      return `${height}px ${width / 2}px 0 ${width / 2}px`
    case 'right':
    default:
      return `${height / 2}px 0 ${height / 2}px ${width}px`
  }
}

// needed for border-color
const reverseDirection = {
  left: 'Right',
  right: 'Left',
  top: 'Bottom',
  bottom: 'Top',
}

/**
 * CSS to represent triangle with any pointing direction with an optional background color. Accepts number or px values for height and width.
 *
 * @example
 * // Styles as object usage
 *
 * const styles = {
 *   ...triangle({ pointingDirection: 'right', width: '100px', height: '100px', foregroundColor: 'red' })
 * }
 *
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${triangle({ pointingDirection: 'right', width: '100px', height: '100px', foregroundColor: 'red' })}
 *
 *
 * // CSS as JS Output
 *
 * div: {
 *  'borderColor': 'transparent',
 *  'borderLeftColor': 'red !important',
 *  'borderStyle': 'solid',
 *  'borderWidth': '50px 0 50px 100px',
 *  'height': '0',
 *  'width': '0',
 * }
 */

function triangle(config: TriangleArgs) {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      messageHandlers('mixins/triangle.js', {
        // eslint-disable-next-line prefer-rest-params
        arrityCheck: { args: arguments, exactly: 1, throw: true },
        typeChecks: {
          param: config,
          type: 'object',
          throw: true,
          required: 'requires a config object as its only parameter. However, you did not provide one.',
        },
      })
    ) {
      return {}
    }
  }

  const {
    pointingDirection,
    height,
    width,
    foregroundColor,
    backgroundColor = 'transparent',
  } = config

  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      !typeChecks('mixins/triangle.js', [
        // TODO: Needs to be a proper enumberable
        {
          param: pointingDirection,
          type: 'string',
          required: 'expects a value(string) for pointingDirection. However, you did not provide one.',
        },
        {
          param: height,
          type: 'string',
          required: 'expects a value(string) for height. However, you did not provide one.',
        },
        {
          param: width,
          type: 'string',
          required: 'expects a value(string) for width. However, you did not provide one.',
        },
        {
          param: foregroundColor,
          type: 'string',
          required: 'expects a value(string) for foregroundColor. However, you did not provide one.',
        },
        { param: backgroundColor, type: 'string' },
      ])
    ) {
      return {}
    }
  }

  const unitlessHeight = parseFloat(height)
  const unitlessWidth = parseFloat(width)

  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      !typeChecks('mixins/triangle.js', [
        {
          param: unitlessHeight,
          type: 'number',
          required: 'requires a pixel based value for height.',
          throw: true,
        },
        {
          param: unitlessWidth,
          type: 'number',
          required: 'requires a pixel based value for width.',
          throw: true,
        },
      ])
    ) {
      return {}
    }
  }

  return {
    borderColor: backgroundColor,
    width: '0',
    height: '0',
    borderWidth: getBorderWidth(
      pointingDirection,
      unitlessHeight,
      unitlessWidth,
    ),
    borderStyle: 'solid',
    /*
    * javascript Object sorting orders 'border-color' after 'border-bottom-color'
    * (bottom-b) is before (bottom-c) - !important is needed
    * { border-bottom-color: 'red', border-color: 'transparent' }
    */
    [`border${reverseDirection[pointingDirection]}Color`]: `${foregroundColor} !important`,
  }
}

export default triangle
