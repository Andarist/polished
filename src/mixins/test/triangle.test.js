// @flow
import triangle from '../triangle'

describe('triangle', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should generate a proper triangle when passed all parameters', () => {
    expect({
      ...triangle({
        foregroundColor: 'red',
        backgroundColor: 'black',
        pointingDirection: 'right',
        height: '10px',
        width: '20px',
      }),
    }).toMatchSnapshot()
  })

  it('should default to a transparent background when not passed a backgroundColor', () => {
    expect({
      ...triangle({
        foregroundColor: 'red',
        pointingDirection: 'right',
        height: '10px',
        width: '20px',
      }),
    }).toMatchSnapshot()
  })

  it('should generate a proper triangle when passed string values for height and width', () => {
    expect({
      ...triangle({
        foregroundColor: 'red',
        backgroundColor: 'black',
        pointingDirection: 'right',
        height: '10px',
        width: '20px',
      }),
    }).toMatchSnapshot()
  })

  it('should properly render top pointing arrow with green foregroundColor, width of 20px and height 20px', () => {
    expect({
      ...triangle({
        foregroundColor: 'green',
        pointingDirection: 'top',
        height: '20px',
        width: '20px',
      }),
    }).toMatchSnapshot()
  })

  it('should properly render right pointing arrow with width of 20px and height 10px', () => {
    expect({
      ...triangle({
        foregroundColor: 'red',
        pointingDirection: 'right',
        height: '10px',
        width: '20px',
      }),
    }).toMatchSnapshot()
  })

  it('should properly render bottom pointing arrow with red foregroundColor, width of 20px and height 20px', () => {
    expect({
      ...triangle({
        foregroundColor: 'red',
        pointingDirection: 'bottom',
        height: '20px',
        width: '10px',
      }),
    }).toMatchSnapshot()
  })

  it('should properly render left pointing arrow with blue foregroundColor, width of 10px and height 20px', () => {
    expect({
      ...triangle({
        foregroundColor: 'blue',
        pointingDirection: 'left',
        height: '20px',
        width: '10px',
      }),
    }).toMatchSnapshot()
  })

  it('should throw an error when not passed any parameters', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    triangle()
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not passed a foregroundColor', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    triangle({
      pointingDirection: 'left',
      height: '20px',
      width: '10px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not passed a pointingDirection', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    triangle({
      foregroundColor: 'blue',
      height: '20px',
      width: '10px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not passed a height', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    triangle({
      foregroundColor: 'blue',
      pointingDirection: 'left',
      width: '10px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not passed a width', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    triangle({
      foregroundColor: 'blue',
      pointingDirection: 'left',
      height: '20px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for foregroundColor', () => {
    triangle({
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      foregroundColor: true,
      pointingDirection: 'left',
      height: 'inherit',
      width: '10px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for pointingDirection', () => {
    triangle({
      foregroundColor: 'blue',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      pointingDirection: true,
      height: 'inherit',
      width: '10px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for height', () => {
    triangle({
      foregroundColor: 'blue',
      pointingDirection: 'left',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      height: 10,
      width: '10px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for width', () => {
    triangle({
      foregroundColor: 'blue',
      pointingDirection: 'left',
      height: '10px',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      width: 10,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed an invalid height', () => {
    triangle({
      foregroundColor: 'blue',
      pointingDirection: 'left',
      height: 'inherit',
      width: '10px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed an invalid width', () => {
    triangle({
      foregroundColor: 'blue',
      pointingDirection: 'left',
      height: 'inherit',
      width: '10px',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
})
