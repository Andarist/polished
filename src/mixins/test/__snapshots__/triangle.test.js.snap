exports[`triangle should default to a transparent background when not passed a backgroundColor 1`] = `
Object {
  "borderColor": "transparent",
  "borderLeftColor": "red !important",
  "borderStyle": "solid",
  "borderWidth": "5px 0 5px 20px",
  "height": "0",
  "width": "0",
}
`;

exports[`triangle should generate a proper triangle when passed all parameters 1`] = `
Object {
  "borderColor": "black",
  "borderLeftColor": "red !important",
  "borderStyle": "solid",
  "borderWidth": "5px 0 5px 20px",
  "height": "0",
  "width": "0",
}
`;

exports[`triangle should generate a proper triangle when passed string values for height and width 1`] = `
Object {
  "borderColor": "black",
  "borderLeftColor": "red !important",
  "borderStyle": "solid",
  "borderWidth": "5px 0 5px 20px",
  "height": "0",
  "width": "0",
}
`;

exports[`triangle should properly render bottom pointing arrow with red foregroundColor, width of 20px and height 20px 1`] = `
Object {
  "borderColor": "transparent",
  "borderStyle": "solid",
  "borderTopColor": "red !important",
  "borderWidth": "20px 5px 0 5px",
  "height": "0",
  "width": "0",
}
`;

exports[`triangle should properly render left pointing arrow with blue foregroundColor, width of 10px and height 20px 1`] = `
Object {
  "borderColor": "transparent",
  "borderRightColor": "blue !important",
  "borderStyle": "solid",
  "borderWidth": "10px 10px 10px 0",
  "height": "0",
  "width": "0",
}
`;

exports[`triangle should properly render right pointing arrow with width of 20px and height 10px 1`] = `
Object {
  "borderColor": "transparent",
  "borderLeftColor": "red !important",
  "borderStyle": "solid",
  "borderWidth": "5px 0 5px 20px",
  "height": "0",
  "width": "0",
}
`;

exports[`triangle should properly render top pointing arrow with green foregroundColor, width of 20px and height 20px 1`] = `
Object {
  "borderBottomColor": "green !important",
  "borderColor": "transparent",
  "borderStyle": "solid",
  "borderWidth": "0 10px 20px 10px",
  "height": "0",
  "width": "0",
}
`;

exports[`triangle should throw a warning when passed more than one parameter 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/triangle.js --%c

triangle %cexpects 1 parameter. However, you passed 2 parameters. These additional parameters were ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when not passed a foregroundColor 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a value(string) for foregroundColor. However, you did not provide one.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when not passed a height 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a value(string) for height. However, you did not provide one.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when not passed a pointingDirection 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a value(string) for pointingDirection. However, you did not provide one.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when not passed a width 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a value(string) for width. However, you did not provide one.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when not passed any parameters 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects 1 parameter. However, you only passed 0 parameters. Please provide 1 additional parameter.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %crequires a config object as its only parameter. However, you did not provide one.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when passed a non-object value 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a parameter of type object. However, you passed 1(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when passed a non-string value for foregroundColor 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a 4th parameter of type string. However, you passed true(boolean) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when passed a non-string value for height 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a 2nd parameter of type string. However, you passed 10(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when passed a non-string value for pointingDirection 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a 1st parameter of type string. However, you passed true(boolean) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when passed a non-string value for width 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %cexpects a 3rd parameter of type string. However, you passed 10(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when passed an invalid height 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %crequires a pixel based value for height.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`triangle should throw an error when passed an invalid width 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/triangle.js --%c

triangle %crequires a pixel based value for height.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#triangle%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
